Dev

manualcheg opened the pull request:
Спринт 14
Реализован функционал аудиоплеера
Реализован поиск с задержкой без нажатия кнопок
Добавление ProgressBar при API-запросах
Перевод запросов к itunes на http

manualcheg committed:
Спринт 14
Реализация функциональности плеера
Реализация поиска с задержкой без нажатия на кнопки
Добавка ProgressBar для запросов к API
Перевод на http-запросы к itunes

manualcheg committed:
Спринт 14
Реализация функциональности плеера
Реализация поиска с задержкой без нажатия на кнопки
Добавка ProgressBar для запросов к API
Перевод на http-запросы к itunes
Плюс Починка логики пряток списка прослушанных треков

kirillNay commented on app/src/main/java/com/practicum/playlistmaker/PlayerActivity.kt:
```diff
@@ -23,21 +27,23 @@ class PlayerActivity : AppCompatActivity() {
+    private lateinit var playbackTime: TextView
+
+    private var mediaPlayer = MediaPlayer()
+    private var mainThreadHandler: Handler? = null
```
⏫ `mainThreadHandler` можно инициализировать сразу при создании класса. `Handler` и `Looper` не зависят от жизненного цикла какого-либо экрана и доступны в любой части приложения

manualcheg replied:
Понял. Спасибо! 
Сделал

kirillNay commented on app/src/main/java/com/practicum/playlistmaker/PlayerActivity.kt:
```diff
@@ -48,7 +54,7 @@ class PlayerActivity : AppCompatActivity() {
         val primaryGenreName = intent.getStringExtra("primaryGenreName")
         val country = intent.getStringExtra("country")
-
+        val previewUrl = intent.getStringExtra("previewUrl")
```
⏫ Обычно, когда возникает необходимость передавать некий объект через `Intent`, сохранять объект в `onSavedInstanceState` или сохранять объект в аргументах `Fragment`'ов (про фрагменты будет рассказано в следующих спринтах) андроид использует `Bundle` для сохранения и передачи данных. Особенность `Bundle` заключается в том, что он представляет собой коллекцию пар *"ключ-значение"*, где ключи являются строками, а значения могут быть различных типов данных, таких как строки, числа, а также объекты классов `Parcelable` и `Serializable`, то есть сериализуемые объекты.

В Android существует несколько способов сериализации объектов:
`Serializable`, `Parcelable`, `JSON`.

В угоду масштабируемости и гибкости гораздо лучше воспользоваться одним из этих способ сериализации объектов, чем передавать поля по-отдельности.

manualcheg replied:
Передал объект в интент через сериализацию Gson().
Спасибо!

kirillNay commented on app/src/main/java/com/practicum/playlistmaker/PlayerActivity.kt:
```diff
@@ -23,21 +27,23 @@ class PlayerActivity : AppCompatActivity() {
-        group = findViewById(R.id.player_group_album_visibility)
+        setViews()
+
+        val startOfDataExpression = 0
```
⏫ Лучше эти значения вынести в константы - `const val` внутри `companion object`

manualcheg replied:
OK

kirillNay commented on app/src/main/java/com/practicum/playlistmaker/PlayerActivity.kt:
```diff
@@ -63,20 +69,112 @@ class PlayerActivity : AppCompatActivity() {
+        buttonPlay.setOnClickListener {
+            val newThread = Thread {
+//                looper = Looper()
+                playbackControl()
```
⏫ Нет необходимости для `playbackControl()` создавать объект Thread и вызывать в отдельном потоке.

manualcheg replied:
Уже понял это, перечитав часть теории про MediaPlayer. Это была попытка предотвратить постоянный рестарт прокрутки имени исполнителя и названия трека.
Убрал.

kirillNay commented on app/src/main/java/com/practicum/playlistmaker/PlayerActivity.kt:
```diff
@@ -63,20 +69,112 @@ class PlayerActivity : AppCompatActivity() {
+        private const val STATE_PLAYING = 2
+        private const val STATE_PAUSED = 3
+        private const val PLAYBACK_TIME_RENEW_DELAY = 300L
+    }
```
⏫ В названии константных переменных, лучше указывать название величины. В данном случае `PLAYBACK_TIME_RENEW_DELAY_MS`

manualcheg replied:
Изменил. Постараюсь запомнить.

kirillNay commented on app/src/main/res/drawable/image_pause_button.xml:
```diff
@@ -0,0 +1,9 @@
+    android:viewportHeight="84">
+  <path
+      android:pathData="M42,0.333C19,0.333 0.333,19 0.333,42C0.333,65 19,83.667 42,83.667C65,83.667 83.667,65 83.667,42C83.667,19 65,0.333 42,0.333ZM37.833,58.667H29.5V25.333H37.833V58.667ZM54.5,58.667H46.167V25.333H54.5V58.667Z"
+      android:fillColor="#1A1B22"/>
```
⏫ Цвета в векторных изображениях также можно выносить в файл ресурсов. Так, можно организовать более гибкую дизайн-систему приложения.
Кроме того, это поможет экономить занимаемое ресурсами место, если нам нужно иметь иконки разных цветов для разных тем приложения. Ведь вместо того, чтобы создавать две иконки и класть в `drawable` и `drawable-night`, мы можем создать одну иконку в `drawable`, а файл ресурса с цветом будет задан в `values` и `values-night`

manualcheg replied:
Понял. У кнопки Play сделал так.
У кнопки Pause уже была такая реализация, но через цвет светлой/темной тем.

kirillNay commented on app/src/main/java/com/practicum/playlistmaker/TrackItemAdapter.kt:
```diff
@@ -36,6 +36,7 @@ class TrackItemAdapter(
             intent.putExtra("releaseDate",track.releaseDate)
```
⚠️ Не был добавлен функционал `debounce'а` при нажатии на элемент трека. 
Советую также вынести обработку нажатия в `Activity`, чтобы иметь флаг для дебаунса общий для всех элементов

manualcheg replied:
Проглядел.
Добавил.

manualcheg replied:
Только как вынести обработку нажатия, увы, не знаю.

manualcheg committed:
Спринт 14
Добавка пропущенного дебаунса нажатия на элемента списка
Доработки в соответствии с замечаниями ревьюера

manualcheg committed:
Спринт 14
Чистка замечаний IDE
